lib/clients/AliClient.py

```diff
@@ -1,31 +1,115 @@
+from aliyunsdkcore.request import CommonRequest
+from aliyunsdkcore.client import AcsClient
+from ..models.Snapshot import Snapshot
+from ..models.Volume import Volume
+from ..models.Attachment import Attachment
+import json
 
 class AliClient(BaseClient):
     def __init__(self, operation_name, configuration, directory_persistent, directory_work_list, poll_delay_time,
                  poll_maximum_time):
         super(AliClient, self).__init__(operation_name, configuration, directory_persistent, directory_work_list,
                                         poll_delay_time, poll_maximum_time)
         if configuration['credhub_url'] is None:
-            auth = oss2.Auth(configuration['access_key_id'], configuration['secret_access_key'])
-            endpoint = configuration['endpoint']
+            self.__setCredentials(
+                configuration['access_key_id'], configuration['secret_access_key'], configuration['region_name'])
         else:
             credentials = self._get_credentials_from_credhub(configuration)
-            auth = oss2.Auth(credentials['access_key_id'], credentials['secret_access_key'])
-            endpoint = credentials['endpoint']
+            self.__setCredentials(
+                credentials['access_key_id'], credentials['secret_access_key'], credentials['region_name'])
+        self.endpoint = configuration['endpoint']
+        # +-> Create compute and storage clients
+        self.compute_client = self.create_compute_client()
+        self.storage_client = self.create_storage_client()
+
         # +-> Check whether the given container exists
-        self.container = self.get_container(auth, endpoint)
+        self.container = self.get_container()
         if not self.container:
             msg = 'Could not find or access the given container.'
             self.last_operation(msg, 'failed')
             raise Exception(msg)
+        
+        # skipping some actions for blob operation
+        if operation_name != 'blob_operation':
+            # +-> Get the availability zone of the instance
+            self.availability_zone = self._get_availability_zone_of_server(
+                configuration['instance_id'])
+            if not self.availability_zone:
+                msg = 'Could not retrieve the availability zone of the instance.'
+                self.last_operation(msg, 'failed')
+                raise Exception(msg)
+
+    def __setCredentials(self, access_key_id, secret_access_key, region_name):
+        self.__aliCredentials = {
+            'access_key_id': access_key_id,
+            'secret_access_key': secret_access_key,
+            'region_name': region_name
+        }
+
+    def create_compute_client(self):
+        try:
+            credentials = self.__aliCredentials
+            compute_client = AcsClient(credentials['access_key_id'], credentials['secret_access_key'], credentials['region_name'], auto_retry=True,
+            max_retry_time=10, timeout=30)
```

Comment:
Is exponential backoff available here as retry mechanism?

Reply:
Yes, it is part of the client